Create separate build directories based on version

cadmic opened the pull request:
This adds a `VERSION` make variable, where each version has separate build directories, output ROMs, and compiler flags. So:

* `make` creates `build/gc-eu-mq-dbg` and `oot-gc-eu-mq-dbg.z64`
* `make VERSION=gc-eu-mq` creates `build/gc-eu-mq` and `oot-gc-eu-mq.z64`

`make clean` also uses `VERSION` to decide which directory to clean. However, `make setup` still only extracts assets from the debug ROM (`baserom.z64`), which I think is good enough for `gc-eu-mq`.

It's still not clear how exactly to do multi-version support, but in any case I think we'll want parallel build directories like this. For example, when making a change for one version, it's very nice to be able to quickly rebuild the debug ROM to check if it's still OK rather than doing a clean and full rebuild, and then another clean and rebuild to get back to what you were doing.

I wasn't sure what to do with the `spec` file, where every include path starts with `build/`. I changed this to `$(BUILD_DIR)/`, taking inspiration from the N64 SDK examples where they often use `$(ROOT)/` for include paths (although here it's implemented using `sed`). Another option would be to remove the `build/` prefix entirely, but I'm not sure if that would break modding workflows or something, if you wanted to include stuff from outside the build directory.

I tried to keep the existing tools working, but not all of them know about multiple versions for now. There are some tools that are so old that I couldn't figure out how to test them, so I deleted them instead in #1590.

cadmic commented:
`find_unused_asm.sh` is failing in CI, but I don't know how to fix it because it also fails for me locally on `main`. Could someone take a look?

jpburnett commented:
build is failing because jenkinsfile depends on the `find_unused_asm.sh` script you deleted.

Dragorn421 commented on Makefile:
```diff
@@ -298,84 +311,84 @@ $(O_FILES): | asset_files
-build/$(SPEC): $(SPEC)
-	$(CPP) $(CPPFLAGS) $< > $@
+$(BUILD_DIR)/$(SPEC): $(SPEC)
+	$(CPP) $(CPPFLAGS) $< | sed -e 's|$$(BUILD_DIR)|$(BUILD_DIR)|g' > $@
```
I think it would be cleaner to use cpp (the C preprocessor) instead, for example

```suggestion
	$(CPP) $(CPPFLAGS) '-DBUILD_DIR=$(BUILD_DIR)' $< > $@
```
and use `BUILD_DIR` instead of `$(BUILD_DIR)` in the spec

Wait for more input to make this change though. There are alternatives like having the spec use `build/VERSION/` which may be discussed too

cadmic replied:
CPP does not replace macros in string literals 🙁

Dragorn421 replied:
Oh right the spec format is `include "path/to/tu.o"`... I didn't realize that cpp would see it as a string
Bleh
(ftr I'm a long time hater of the spec format <img src="https://cdn.discordapp.com/emojis/943102210366119956.webp?size=56&quality=lossless" alt="skawoUHHUH" height="24"> )

Dragorn421 reviewed (commented):
[Thinking aloud]
Given how many things use the build/ folder (I mean build/$VERSION/) I think it would be appropriate to rely on an environment variable to have tools automatically use it. And since it depends on the oot version maybe the oot version should be meant as an environment variable too. Like:
```sh
export OOT_VERSION=gc-eu-mq-dbg
export OOT_BUILD_DIR=build/$OOT_VERSION  # or somewhere else
```
(using an oot_ prefix to prevent conflicts)
(the makefile would just inherit those from the environment too:
```mk
# note: using ?=
OOT_VERSION ?= gc-eu-mq-dbg
OOT_BUILD_DIR ?= build/$(OOT_VERSION)
```
then just `make` (no need to set those)
)

cadmic commented:
The `find_unused_asm.sh` check didn't seem useful anyway so I deleted it, let me know if you object

fig02 commented:
`find_unused_asm.sh` was helpful when we were decompiling and needed to manually remove asm files when we matched a file.
I guess its fine to remove, but how are we planning to handle asm for multi-version support as we work on it? 

We may not have an answer to this question now. But if we take the same approach of asm in the repo, this script may find a use again.

cadmic commented:
I kept the script around, but only deleted it from the Jenkinsfile, because I wasn't sure how to fix it properly (it seems like it depends on the state of the previous build) and because I think there's basically zero chance this check will detect a real issue now

fig02 commented:
Okay. Yeah it wont be relevant again until we start decompiling the other versions, and even then the approach we take may not be the same. It simply makes sure that .s file for decompiled functions are removed.

fig02 commented on Makefile:
```diff
@@ -14,6 +14,10 @@ NON_MATCHING ?= 0
 ORIG_COMPILER ?= 0
 # If COMPILER is "gcc", compile with GCC instead of IDO.
 COMPILER ?= ido
+# Target game version. Currently the following versions are supported:
+#   gc-eu-mq       GameCube Europe/PAL Master Quest
```
Id remove the non debug comment until we actually support it

cadmic replied:
Yeah, I guess I meant "supported by the Makefile" but it was confusing. I clarified that gc-eu-mq is a WIP

fig02 reviewed (approved):
Im approving to say I agree with the idea of separate build directories, but I have nothing to add about implementation. I defer that to the usual tooling/makefile experts.
So, ideally, would like 2 other approvals on top of mine.

cadmic commented:
@Dragorn421 

An environment variable is an interesting idea, but I don't know how useful it would be just yet. Maybe we can revisit once more people are using multiple versions?

(FWIW I don't think `VERSION` would conflict with anything, I googled for a bit and couldn't find any programs that use it)

Dragorn421 commented on Makefile:
```diff
@@ -41,7 +46,20 @@ ifeq ($(NON_MATCHING),1)
 endif
 
+# Version-specific settings
+ifeq ($(VERSION),gc-eu-mq)
+  CFLAGS += -DNON_MATCHING -DNDEBUG
+  CPPFLAGS += -DNON_MATCHING -DNDEBUG
+  OPTFLAGS := -O2 -g3
+  COMPARE := 0
```
The "build a non matching rom at first and work towards OK" approach is new (to us at least), and I don't think it's been properly discussed and agreed upon

Plus idk if gc-eu-mq is the version we want to do next. gc-eu-dbg-* are other candidates

Or I missed both discussions of the above two points.

In any case I think this PR can solely focus on moving `build/` to `build/gc-eu-mq-dbg/` (in a way that involves a version variable ofc)

cadmic replied:
That's fair, I removed all mentions of gc-eu-mq.

cadmic commented:
> I think if we were to implement this idea this would be the adequate PR since it already touches everything that relies on / touches build/

idk, we'll have to touch them all again anyway to give them proper version support. I intentionally didn't do that here because this PR is already has pretty major changes (like the spec format and rom names).

I guess I'm still not sold on the environment variable. I've been switching versions a lot, since for basically every change I also want to check the debug ROM still matches. But environment variables seem more suited for things that probably will never change during a terminal session, like `TZ`, `PATH`, or `EDITOR`. It could be pretty confusing if you ran `diff.py` expecting to diff one version, but it's actually a different version, and there's nothing in the command line to reflect that.

fig02 commented:
Im not totally set on the environment variable idea rn either, but could see it happening in the future as well

Dragorn421 commented:
Should this PR add a `--version` option to every tool instead of hardcoding the version? Similarly to how it is introduced as a variable in the Makefile. I don't think there's a reason to postpone such a change, we will for sure need it and this PR touches all the relevant spots

Also I acknowledge the baserom is kept expected as being gc-eu-mq-dbg for now, without any preparation for "multiversion baserom" (it's still named `baserom.z64`), it can be done separately

cadmic commented:
Okay, I went through all the tools again:

* `diff.py`: added a version flag for object file paths and make options. Elf diffing and ROM diffing probably can't work with my proposed setup though
* `sym_info.py`: added a version flag
* `first_diff.py`: already has a version flag, but is broken in my proposed setup because expected/z64.map does not exist. I plan on fixing this but I'd like to postpone it for now
* `format.py`: IMO it's fine to only consider the debug ROM (the only thing it affects is which generated headers to include)
* `assist.py`: TBH I'm really not sure how to make this work. It uses build/z64.map but parses symbols from the baserom? I don't think that will work if we have a non-matching build or a compressed baserom
* `progress.py`: Not relevant in my proposed setup (or needs to be rewritten)
* `find_unused_asm.sh`: Not relevant in my proposed setup

cadmic commented:
@fig02 could you suggest a third reviewer for this PR?

fig02 commented:
I dont really want to call specific people out and make them feel pressured to review. But the people I personally associate with these kinds of tooling/makefile changes are dragorn, tharo, anghelo, roman to name a few (at least among those currently active).
They should only review if they have time and are able :)

jpburnett commented on Jenkinsfile:
```diff
@@ -4,11 +4,6 @@ pipeline {
     }
 
     stages {
-        stage('Check for unused asm') {
-            steps {
-                sh './tools/find_unused_asm.sh'
-            }
-        }
```
While this is removed, what was going to be the plan going forward for handling asm in other versions? Is the plan to go forward with transient asm?

cadmic replied:
I'm not sure yet, but if we do commit asm again, we can always add a similar check back. This check serves no purpose now though

Dragorn421 replied:
Yeah we're going with transient asm, though it's not 100% decided I don't see a path where transient asm can't be made to work

Dragorn421 commented on format.py:
```diff
@@ -29,7 +29,7 @@
 # Compiler options used with Clang-Tidy
 # Normal warnings are disabled with -Wno-everything to focus only on tidying
-INCLUDES = "-Iinclude -Isrc -Ibuild -I."
+INCLUDES = "-Iinclude -Isrc -Ibuild/gc-eu-mq-dbg -I."
```
Unfortunately always including build/ of a single version means you expect that version to be built which may not be the case (e.g. you may be building a different version than this one without having ever built the version used for format)

Note I am not sure if it is even relevant to include build/ for formatting, afaik the only reason for including build/ in general is for assets including .inc.c files, and the formatter isn't run on assets, but I may be forgetting stuff

cadmic replied:
Hmm it seems to work find even if this directory does not exist. Should I remove this include? (What about in the `.vscode` docs?)

Dragorn421 replied:
For vscode settings it's useful because it makes the ide find those .inc.c files

I'm not sure what to do about formatting (I think the script silences all errors btw?)
I'll look into it and report back

Roman971 replied:
Iirc clang-tidy needs proper includes otherwise it will error instead of doing what it needs to with the file

cadmic replied:
What about adding a check in format.py to fail if build/gc-eu-mq-dbg does not exist? I kinda think gc-eu-mq-dbg should be the "default" version for formatting, since it should be standardized but maybe clang-tidy might make different decisions per-version (though I'm not sure why it would)

Dragorn421 replied:
this is only an issue for the purpose of using the repo tools for different versions, which we postponed to another PR, so we can sweep this under the rug for now and be content with hardcoding gc-eu-mq-dbg for formatting

cadmic replied:
Alright, I made a note to myself about this

cadmic commented on diff_settings.py:
```diff
@@ -1,6 +1,10 @@
+def add_custom_arguments(parser):
+    parser.add_argument("-v", "--version", help="Game version to diff", default="gc-eu-mq-dbg")
```
This is why I wanted to make this change in a separate PR 🙃 

I changed the log form to `--oot-version`, but I don't want to give up on `-v`. I strongly feel that it's important for this argument to be short and intuitive since we'll be typing it a lot, and I don't think any other letter would work. If some tool does someday get a `--verbose` option, I'm okay not having a short option for that, and I don't think people would confuse this for asm-differ's `-V`.

Dragorn421 replied:
I understand wanting to keep -v short but I feel it's a bit useless when you have to type the long "gc-eu-mq-dbg" or similar right after, so I don't think you would type this over and over again anyway, and would rather copypaste it

(personally I would probably end up using shell aliases or wrapper scripts)

I don't think with introducing this as `-v` we would ever move it as something else if a verbose flag was added somewhere. This is the kind of thing you cannot change, because people will be used to it.

Regardless I'm okay with it as -v as things are currently, I'm only resisting because I know it would be much harder to change later

Roman971 replied:
I think for scripts like `diff.py` or `sym_info.py`/`first_diff.py`, it would be ideal if it defaults to the last version built so you don't have to provide the version everytime. There are multiple ways to achieve this, like looking at the last modified files in build/ for instance, or tracking the version currently being worked on in a local file (this would be my preferred option since it's easy to reference anywhere)

fig02 replied:
I also agree with the concerns about -v and --version

cadmic replied:
The primary concern here should be usability since this is something people will interact with a lot. It's true that `-v` is often used for "software version" or "verbose" in other programs, which does make it less usable to use `-v` here for something else. But I think that is outweighed by the usability benefits of using `-v`, namely that it's nice to have a short form for such a common option, and there's no other letter that fits.

> I understand wanting to keep -v short but I feel it's a bit useless when you have to type the long "gc-eu-mq-dbg" or similar right after, so I don't think you would type this over and over again anyway, and would rather copypaste it

Well, gc-eu-mq is still much faster for me to type than copy-pasting it, and fortunately "gc-eu-mq-dbg" is the default so I never have to type it.

> I don't think with introducing this as -v we would ever move it as something else if a verbose flag was added somewhere.

Right, I meant in the off change we did add a verbose flag, I would be much happier typing out `--verbose` than giving up `-v` for the OOT version.

> I think for scripts like diff.py or sym_info.py/first_diff.py, it would be ideal if it defaults to the last version built so you don't have to provide the version everytime.

Defaulting to the last version built sounds way too magical IMO. I've been building and diffing gc-eu-mq and gc-eu-mq-dbg at the same time and having some kind of hidden state sounds very confusing. But even then I would still want a short option to switch between the two.

cadmic replied:
(I've already changed `--version` -> `--oot-version` btw, I'm talking about the short option here)

Dragorn421 replied:
What about a shorter long option (alias) like `--ootv` or `--ov`

fig02 replied:
-ov sounds like a fine compromise to me

cadmic replied:
Eh those both feel kinda ugly to me. `--ov` could be confused for `-ov` (i.e. passing both `-o` and `-v`), and `--ootv` makes me think of https://www.twitch.tv/ootv. Not nearly as nice as `-v`, which clearly stands for "version", which in the context of this repo almost always means "OOT game version" and not "software tool version"

Dragorn421 replied:
-v is short for --verbose in [most] command line tools

I really do think it's worth taking into account and adhering to *de facto* standards when designing something new.
It makes less to take in for someone experienced, and is more useful / less misleading to learn for someone inexperienced.

[If I can rant for a bit this is why linux and related software is so good, developments are made to blend in instead of making flashy changes à la Microsoft because more new=more sales]

fig02 replied:
I hear your points and I understand you feel strongly about this, but 3 people so far feel like it needs to change. We can get even more opinions on this if you want, but at the moment its pretty lopsided.

--ov as an alias, with two dashes to signify it being one argument name, still sounds like the best compromise so far

cadmic replied:
I have taken those de facto standards into account, and decided that other factors have outweighed them. The primary goal should be to make a tool that people enjoy using, and adhering to standards is one component of that, but it's a mistake to make that the end goal.

I'm not looking for more opinions, I'm looking for a more convincing argument than "it could be confusing". Maybe this github thread is not the best place for that though, so I'm going to revert adding version flags for this PR and we can revisit it later

fig02 replied:
Its common in this project (and I'd say group projects in general) to concede to changes you may not like for the sake of compromise with the majority opinion. Feel like it needs to be said now while you are new to contributing here, that majority opinion is our main path forward in any heated debate. 
It will not be the last time people disagree with changes you make, unfortunately, but it happens to everyone. Pushing the conversation to later wont really help I don't think.

cadmic replied:
Don't worry, I understand that. I do think it would help though if I had a little more time to think about this and try out some options. In the meantime I don't want to hold up the rest of this PR since it's kind of blocking everything multiversion-related.

I added versions flags to this PR against my better judgement (I kinda knew this would happen) so I'm just trying to put it back. Sorry if it looks like I'm being a baby about it, I don't mean it that way

fig02 replied:
Alright, maybe later is better in this case then. 
And no worries. I just want to make expectations clear now on how we try to operate in situations like this, going forward 😄 Its not fun for anyone to deal with disagreements, but we gotta move forward somehow so we can keep working on the project we like FeelsOkayMan

Dragorn421 replied:
aye <img src="https://cdn.discordapp.com/emojis/982516941375823942.webp?size=56&quality=lossless" alt="OkaygeBusiness" height="24">

Sorry for forcing your hand on making these changes now. Still we're gonna have to figure it out soon!

cadmic commented on Makefile:
```diff
@@ -14,6 +14,9 @@ NON_MATCHING ?= 0
 COMPILER ?= ido
+# Target game version. Currently only the following version is supported:
+#   gc-eu-mq-dbg   GameCube Europe/PAL Master Quest Debug (default)
+VERSION ?= gc-eu-mq-dbg
```
Should I change them all to `:=`? I don't have a preference for `?=` vs `:=` but I think that all of the build options should be consistent.

(again I don't think it's a risk that some system-wide environment variable will be called `VERSION`. If anything, `COMPILER` seems like the most risky to me)

Dragorn421 replied:
Yeah I do think no variable in the makefile has any business being a `?=` and they're all based on making the mistake that "`?=` allows overriding like `make A=...`" (which is wrong, it allows defaulting to environment variables if set)

Leave that for another PR to fix though

cadmic replied:
Okay thanks, created #1599

Dragorn421 commented on tools/find_unused_asm.sh:
This was probably said in discord but, why are we not removing this file right now?

fig02 replied:
At the start of this multiversion conversation we werent sure if it would be relevant again. But if it looks like we are going with cadmics proposal, then it wont be needed

cadmic replied:
Okay, I went ahead and deleted it based on dragorn's comment above:

> Yeah we're going with transient asm, though it's not 100% decided I don't see a path where transient asm can't be made to work

Dragorn421 replied:
It should be mentioned this file is still mentioned in the tutorial, so it's one more thing the tutorial is outdated about. However there are many outdated things in the tutorial anyway so I told cadmic to not worry about the tutorial

fig02 replied:
Much of the tutorial is largely irrelevant cause its about how to decomp the game for the first time. Our procedures are going to be changing drastically from here on

Dragorn421 commented on docs/tutorial/data.md:
````diff
@@ -536,28 +536,28 @@ Oh no! What went wrong?
 ```
 $ ./firstdiff.py
-First difference at ROM addr 0x144F4, gDmaDataTable (RAM 0x80016DA0, ROM 0x12F70, build/asm/dmadata.o)
+First difference at ROM addr 0x144F4, gDmaDataTable (RAM 0x80016DA0, ROM 0x12F70, build/gc-eu-mq-dbg/asm/dmadata.o)
 Bytes: 00:E3:F9:D0 vs 00:E3:F9:E0
-Instruction difference at ROM addr 0xE3ED48, En_Jj_InitVars (RAM 0x80A88140, ROM 0xE3ED10, build/src/overlays/actors/ovl_En_Jj/z_en_jj.o)
+Instruction difference at ROM addr 0xE3ED48, En_Jj_InitVars (RAM 0x80A88140, ROM 0xE3ED10, build/gc-eu-mq-dbg/src/overlays/actors/ovl_En_Jj/z_en_jj.o)
 Bytes: 40:00:00:00 vs 00:F0:00:00
````
I appreciate wanting to update the firstdiff.py output but it's anachronistic here, the output of the tool changed beyond just the build/ path since this was copypasted. I would just leave what the tutorial currently has for first_diff.py output (same below)

(changes in the tutorial about spec segments are fine, I mean specifically the firstdiff.py stuff)

cadmic replied:
Okay, I reverted the output

Dragorn421 reviewed (approved):
Let's go!

I recommend deleting `build`, `expected` and the built elf/rom to reclaim ~500MB of wasted storage space once you stop using branches where this isn't yet merged:
`rm -rf build expected zelda_ocarina_mq_dbg.elf zelda_ocarina_mq_dbg.z64`

Yanis42 reviewed (approved):
works on my machine™ (I tried building, make clean/distclean and the scripts changed in this PR except ``tools/assist.py`` because I don't know how to use it)

Thar0 commented on Makefile:
```diff
@@ -298,84 +305,84 @@ $(O_FILES): | asset_files
-build/$(SPEC): $(SPEC)
-	$(CPP) $(CPPFLAGS) $< > $@
+$(BUILD_DIR)/$(SPEC): $(SPEC)
+	$(CPP) $(CPPFLAGS) $< | sed -e 's|$$(BUILD_DIR)|$(BUILD_DIR)|g' > $@
```
```suggestion
	$(CPP) $(CPPFLAGS) $< | $(SPEC_REPLACE_TOKENS) > $@
```
`SPEC_REPLACE_TOKENS := sed -e 's|$$(BUILD_DIR)|$(BUILD_DIR)|g'`
Does something like this work? I think it could be nice to use this to consolidate the two cases where this is used.
(Also if you think there's a better name feel free to use it)

cadmic replied:
Okay, I made `SPEC_REPLACE_VARS`. I didn't go all the way because I think it's a bit strange to write a file when reading the Makefile instead of while executing it

Thar0 reviewed (commented):
One question/suggestion

Dragorn421 commented:
> I recommend deleting `build`, `expected` and the built elf/rom to reclaim ~500MB of wasted storage space once you stop using branches where this isn't yet merged: `rm -rf build expected zelda_ocarina_mq_dbg.elf zelda_ocarina_mq_dbg.z64`

Bump!
